### PR TITLE
rememberable and timeoutable should work together

### LINKED ISSUE #4778: timeout and rememberable
Why was the timeout and rememberable compatibility removed?

```ruby
def timedout?(last_access)
    return false if remember_exists_and_not_expired?
```

https://github.com/plataformatec/devise/blob/master/lib/devise/models/timeoutable.rb#L30
(removed in commit https://github.com/plataformatec/devise/commit/4ec7dc0f2777bde10125dcf6f72c536232086665)

## Current behavior

Timeout and Rememberable do not work together. The shorter one overrules the longer one.

## Expected behavior

I would expect rememberable to control sessions via cookie, while timeoutable can still control session time without cookies. (This is both relevant for http, where cookies are disabled on `secure: true`, as well as for users who explicitly choose not to select remember me).

At the moment, when I want to use rememberable, I have not control what so ever on session time, just on the session cookie expiration time.

### PR DESCRIPTION
**Pull request by prithvi16**
Fixes #4778

Now if rememberable is enabled and has not expired then `timedout?` returns false. 
Added a test for this behavior.

### CODE REVIEW

File: lib/devise/models/rememberable.rb [hunk 1]
```diff
@@ -100,6 +100,10 @@ def rememberable_options
       def after_remembered
       end
 
+      def remember_expired?
+        remember_created_at.nil? || (remember_expires_at <= Time.now.utc)
```
Comment (duderamos) [inline lib/devise/models/rememberable.rb]: Shouldn't we consider app timezone?

```ruby
Time.zone.now
```

**Comment by wmene** [comment]
Any idea when this will get merged?

File: lib/devise/models/rememberable.rb (hunk 1 above)
Comment (prithvi16) [reply lib/devise/models/rememberable.rb]: @duderamos yes, makes sense. Making this change.

File: lib/devise/models/timeoutable.rb [hunk 2]
```diff
@@ -27,10 +27,16 @@ def self.required_fields(klass)
...
+      # Returns false if Rememberable has been enabled and has not expired
       def timedout?(last_access)
+        return false if remember_exists_and_not_expired?
         !timeout_in.nil? && last_access && last_access <= timeout_in.ago
       end
 
+      def remember_exists_and_not_expired?
+        return false unless respond_to?(:remember_created_at) && respond_to?(:remember_expired?)
+        remember_created_at && !remember_expired?
+      end
```
Comment (feliperenan) [inline lib/devise/models/timeoutable.rb]: Let's leave a line blank between these two methods 🎨

File: test/models/timeoutable_test.rb [hunk 3]
```diff
@@ -50,4 +50,10 @@ class TimeoutableTest < ActiveSupport::TestCase
     user = create_admin(remember_created_at: Time.current)
     assert user.timedout?(31.minutes.ago)
   end
+
+  test 'should return false if rememberable is enabled and has not expired' do
+    user = create_user
+    user.remember_me!
+    refute user.timedout?(31.minutes.ago)
+  end
 end
```
Comment (feliperenan) [inline test/models/timeoutable_test.rb]: We could add one more test to cover the scenario that the rememberable is enabled and is expired.

File: lib/devise/models/rememberable.rb [hunk 4]
```diff
@@ -100,6 +100,10 @@ def rememberable_options
       def after_remembered
       end
 
+      def remember_expired?
+        remember_created_at.nil? || (remember_expires_at <= Time.zone.now.utc)
```
Comment (duderamos) [inline lib/devise/models/rememberable.rb]: This is not exactly what I meant here: https://github.com/plataformatec/devise/pull/4926#pullrequestreview-155670776. Your change will end up returning the same value as before.

Consider the following:
```suggestion
        remember_created_at.nil? || (remember_expires_at <= Time.zone.now)
```

So the application will check reminders accordingly to local timezone.

**Pull request closed without merging**